pxd_fastpath.c

```diff
@@ -107,36 +108,29 @@ void _generic_start_io_acct(struct request_queue *q, int rw,
 #endif
 #endif
 
-// cached info at px loadtime, to gracefully handle hot plugging cpus
-static int __px_ncpus;
-
 // A private global bio mempool for punting requests bypassing vfs
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,18,0)
 static struct bio_set pxd_bio_set;
 #endif
 #define PXD_MIN_POOL_PAGES (128)
 static struct bio_set* ppxd_bio_set;
 
-// global thread contexts
-static struct thread_context *g_tc;
-
-// forward decl
-static int pxd_io_writer(void *data);
-static int pxd_io_reader(void *data);
 static void __pxd_cleanup_block_io(struct pxd_io_tracker *head);
-static struct pxd_io_tracker* pxd_get_io(struct thread_context *tc, int rw);
-#define pxd_get_writeio(tc)  pxd_get_io(tc, WRITE)
-#define pxd_get_readio(tc)   pxd_get_io(tc, READ)
 
 // congestion callback from kernel writeback module
 int pxd_device_congested(void *data, int bits)
 {
 	struct pxd_device *pxd_dev = data;
 	int ncount = PXD_ACTIVE(pxd_dev);
+	int cpu = get_cpu();
+	struct pcpu_fpstate *statep = per_cpu_ptr(pxd_dev->fp.state, cpu);
+	int suspend = READ_ONCE(statep->suspend);
```

Comment:
Why is it ok to access the percpu without lock here?

Reply:
The primary reason per cpu got introduced is in the IO path, in make_request, when the IO is processed, if its not suspended, is processed straight. If suspended, then it is confirmed under lock and enqueued. So the common operation is always without lock, because suspending IO happens only on congestion, failover etc.

the lock (suspend_lock) is needed to ensure suspend queuing is consistent with suspend request. Here one relies only on the current state, so suspend_lock is not needed. This queueing in IO path is going away in failover patch next cycle.